src/process.c

```diff
@@ -463,59 +466,450 @@ cc_run_hook(struct oci_cfg_hook* hook, const gchar* state,
-			flags,
-			(GSpawnChildSetupFunc)cc_oci_setup_child,
-			(gpointer)config,
-			&config->state.workload_pid,
-			NULL,
-			NULL,
-			NULL,
-			&err);
+	if (! cc_oci_fd_set_cloexec (child_err_pipe[1])) {
+		g_critical ("failed to set close-exec bit on fd");
+		goto out;
+	}
+
+	pid = config->state.workload_pid = fork ();
+	if (pid < 0) {
+		g_critical ("failed to create child: %s",
+				strerror (errno));
+		goto out;
+	}
+
+	if (! pid) {
+		g_autofree gchar  *netcfg = NULL;
+
+		/* child */
+		pid = config->state.workload_pid = getpid ();
+
+		close (net_cfg_pipe[1]);
+		close (child_err_pipe[0]);
+
+		/* block waiting for parent to send network config.
+		 *
+		 * This arrives in 2 parts: first, a size_t representing
+		 * the size of the ASCII name=value data that follows.
+		 */
+
+		/* determine how many bytes to allocate for the network
+		 * config.
+		 */
+		g_debug ("child reading netcfg length");
+
+		bytes = read (net_cfg_pipe[0], &len, sizeof (len));
+		if (bytes < 0) {
+			goto child_failed;
+		}
+
+		g_debug ("allocating %lu bytes for netcfg",
+				(unsigned long int)len);
+
+		netcfg = g_new0 (gchar, 1+len);
+		if (! netcfg) {
+			goto child_failed;
+		}
+
+		g_debug ("reading netcfg from parent");
+
+		bytes = read (net_cfg_pipe[0], netcfg, len);
+		if (ret < 0) {
+			goto child_failed;
+		}
+
+		close (net_cfg_pipe[0]);
+
+		g_debug ("adding netcfg to config");
+
+		if (! cc_oci_vm_netcfg_from_str (config, netcfg)) {
+			goto child_failed;
+		}
+
+
+		g_debug ("building hypervisor command-line");
 
+		// FIXME: add network config bits to following functions:
+		//
+		// - cc_oci_container_state()
+		// - oci_state()
+		// - cc_oci_update_options()
+
+		ret = cc_oci_vm_args_get (config, &args);
+		if (! (ret && args)) {
+			goto child_failed;
+		}
+
+		// FIXME: add netcfg to state file
+		ret = cc_oci_state_file_create (config, timestamp);
+		if (! ret) {
+			g_critical ("failed to recreate state file");
+			goto out;
+		}
+
+		g_debug ("running command:");
+		for (p = args; p && *p; p++) {
+			g_debug ("arg: '%s'", *p);
+		}
+
+		if (! cc_oci_setup_child (config)) {
+			goto child_failed;
+		}
+
+		if (execvp (args[0], args) < 0) {
+			g_critical ("failed to exec child %s: %s",
+					args[0],
+					strerror (errno));
+			abort ();
+		}
+	}
+
+	/* parent */
+
+	g_debug ("child pid is %u", (unsigned)pid);
+
+	close (net_cfg_pipe[0]);
+	net_cfg_pipe[0] = -1;
+
+	close (child_err_pipe[1]);
+	child_err_pipe[1] = -1;
+
+	/* create state file before hooks run */
+	ret = cc_oci_state_file_create (config, timestamp);
 	if (! ret) {
-		g_critical ("failed to spawn child: %s", err->message);
-		g_error_free (err);
+		g_critical ("failed to create state file");
+		goto out;
+	}
+
+	/* If a hook returns a non-zero exit code, then an error
+	 * including the exit code and the stderr is returned to
+	 * the caller and the container is torn down.
+	 */
+	ret = cc_run_hooks (config->oci.hooks.prestart,
```

Comment:
@sameo The reason for doing this is are mostly

1. CNM requirements
2. GSpawnChildSetupFunc  limitations

Here is the logic (convoluted)
1. We call into in docker CNM using the networking prehook to setup the networking for the container
   - The docker prehook as a parameter expects the "PID" of the process has its own namespace
        - This process needs to be the process that will ultimately run the "container", in our case    QEMU (i.e. the child pid)
   - The docker CNM then sets up the network on pre-hook invocation within the namespace
   - When the pre-hook returns the network namespace is all setup

Today we spawn the child, get its pid and invoke the prehook from the parent.

Meanwhile the child is setting itself up (to the point it needs the network info).
Once the parent has the network info (pre-hook returns), the child continues.

The actual QEMU cannot be exec'ed in the child till it receives the network info.

Note: There may be other pre-hooks in the future (e.g storage).

Also the g_spwan child setup pre-func has limitations on what we can and cannot do within it.
Hence we cannot call the pre-hooks from GSpawnChildSetupFunc